A19-tls-credential-API

```diff
@@ -0,0 +1,281 @@
+Title
+----
+* Author(s): yihuazhang
+* Approver: vjpai, markdroth
+* Status: Draft
+* Implemented in:
+* Last updated: September 3, 2018
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/gMrROsizrtc
+
+## Abstract
+This proposal aims to incorporate SPIFFE-based mutual TLS into gRPC C core auth stack. Towards the goal, it proposes a new credential API model that 1) allows flexible addition of new TLS-related features (e.g., credential reload) without breaking the API and 2) supports both synchronous and asynchronous implementations of those features. It then proposes SPIFFE and HTTPS-based TLS credential API's that comply with the new model.
+
+## Background
+
+The current TLS implementation in gRPC C core has the following problems when it comes to support SPIFFE-based mutual TLS.
+
+1: It is originally designed to comply with HTTPS semantics which makes it difficult to be configured to support SPIFFE-based mutual TLS. For instance, the server authorization check in the current implementation is enforced locally by leveraging whitelisted endpoint information embedded in peer certificates while in SPIFFE use case, it may be enforced via an external server authorization check service.
+
+2: It realizes TLS-related features such as credential reload and server authorization check in a synchronous manner. It is undesirable in SPIFFE use case as those operations could involve expensive RPC calls to other services which may block gRPC core threads.
+
+3: The current TLS credential API is inflexible in incorporating new TLS-related features in future as not only does it break the API, it also requires all wrapped languages to update their implementations in a lockstep manner.
+
+This proposal introduces a new credential API model that allows flexible addition of TLS-related features that can be implemented in either a synchronous or asynchronous manner. It then proposes new credential API's for both HTTPS and SPIFFE-based TLS that complies with the API model.
+
+### Related Proposals:
+N/A
+
+## Proposal
+
+The first part of proposal introduces credential options and credential create APIs that will be added to include/grpc/grpc_security.h.
```

Comment:
Add a note that although grpc_tls_https_server_credentials_create, grpc_tls_https_credentials_create, grpc_tls_spiffe_server_credentials_create, grpc_tls_spiffe_credentials_create use the same grpc_tls_credentials_options struct as input, certain configure may not be applicable and will be ignored during credential creation. E.g., grpc_ssl_client_certificate_request_type config is only applicable to https_server_credentials creation.